Python/Product/PythonTools/PythonTools/Intellisense/TaskProvider.cs

```diff
@@ -928,61 +937,97 @@ string sourceFile
             return VSConstants.E_NOTIMPL;
         }
 
-        int IVsTaskItem.OnFilterTask(int fVisible) {
+        public int OnFilterTask(int fVisible) {
             return VSConstants.E_NOTIMPL;
         }
 
-        int IVsTaskItem.SubcategoryIndex(out int pIndex) {
+        public int SubcategoryIndex(out int pIndex) {
             pIndex = 0;
             return VSConstants.E_NOTIMPL;
         }
 
-        int IVsTaskItem.get_Checked(out int pfChecked) {
+        public int get_Checked(out int pfChecked) {
             pfChecked = IsChecked ? 1 : 0;
             return VSConstants.S_OK;
         }
 
-        int IVsTaskItem.get_Priority(VSTASKPRIORITY[] ptpPriority) {
+        public int get_Priority(VSTASKPRIORITY[] ptpPriority) {
             ptpPriority[0] = Priority;
             return VSConstants.S_OK;
         }
 
-        int IVsTaskItem.get_Text(out string pbstrName) {
+        public int get_Text(out string pbstrName) {
             pbstrName = Message;
             return VSConstants.S_OK;
         }
 
-        int IVsTaskItem.put_Checked(int fChecked) {
+        public int put_Checked(int fChecked) {
             if (IsCheckedIsReadOnly) {
                 return VSConstants.E_NOTIMPL;
             }
             IsChecked = (fChecked != 0);
             return VSConstants.S_OK;
         }
 
-        int IVsTaskItem.put_Priority(VSTASKPRIORITY tpPriority) {
+        public int put_Priority(VSTASKPRIORITY tpPriority) {
             if (PriorityIsReadOnly) {
                 return VSConstants.E_NOTIMPL;
             }
             Priority = tpPriority;
             return VSConstants.S_OK;
         }
 
-        int IVsTaskItem.put_Text(string bstrName) {
+        public int put_Text(string bstrName) {
             if (MessageIsReadOnly) {
                 return VSConstants.E_NOTIMPL;
             }
             Message = bstrName;
             return VSConstants.S_OK;
         }
+
+        public int BrowseObject(out object ppObj) {
+            ppObj = null;
+            return VSConstants.E_NOTIMPL;
+        }
+
+        public int get_CustomColumnText(ref Guid guidView, uint iCustomColumnIndex, out string pbstrText) {
+            pbstrText = $"{guidView};{iCustomColumnIndex}";
+            return VSConstants.S_OK;
+        }
+
+        public int put_CustomColumnText(ref Guid guidView, uint iCustomColumnIndex, string bstrText) {
+            return VSConstants.E_NOTIMPL;
+        }
+
+        public int IsCustomColumnReadOnly(ref Guid guidView, uint iCustomColumnIndex, out int pfReadOnly) {
+            pfReadOnly = 1;
+            return VSConstants.S_OK;
+        }
+
+        public int GetHierarchy(out IVsHierarchy ppProject) {
+            if (_projectHierarchyIsNull || ProjectHierarchy != null) {
+                ppProject = ProjectHierarchy;
+            } else if (!string.IsNullOrEmpty(SourceFile)) {
+                ppProject = ProjectHierarchy = _serviceProvider.GetProjectFromFile(SourceFile);
```

Comment:
This caching will break if someone changes `SourceFile` after this is called (its setter is public so it can technically be done). We could change its setter to private or make it a read-only field. `ProjectHierarchy` also does not need to be public.

Reply:
This is very much a "don't do that" situation - it's nice to be able to initialize the optional members of class using the assignment syntax rather than having a huge constructor. But in this case, it's not an optional member, so it can be read only.